### PR TITLE
initial work for proxy observation

### PR DESCRIPTION
**Pull request by fkleuver**
# Pull Request

## 📖 Description

This is an initial stab at implementing proxy-based observation as a flexible opt-in (that is, it can be toggled on multiple levels via flags, but the easiest way is via the `@customElement` decorator). This is a preliminary API of course just to try things out.

~~The integration with `observer-locator` is pretty quick-n-dirty and it doesn't catch on every level just yet. Need to find the right places to replace/trap objects.~~

At least observing array length and index setters seems to work when the observation activates at the right level, so seems to be on the right track.

### 🎫 Issues

Related to (but broader than) https://github.com/aurelia/aurelia/issues/76

## 👩‍💻 Reviewer Notes

- Most file changes have to do with a single parameter change that adds lifecycle to hydration so the scope creation can be intercepted. Majority of changes are actually test files.
- Interesting bits to look at: [`binding-context.ts`](https://github.com/aurelia/aurelia/pull/381/files#diff-fb20f0198f8c451e96b4f6c194154bd2R81), [`observer-locator.ts`](https://github.com/aurelia/aurelia/pull/381/files#diff-41934ef11565bd232010c1e170bb633bR44) and of course [`proxy-observer.ts`](https://github.com/aurelia/aurelia/pull/381/files#diff-629432392ecd37dc6a52c15e75ae57beR1)

Would strongly recommend using VS code PR review tool to weed out the tests and stuff

## 📑 Test Plan

- Mostly planning to reuse existing tests and see if they all pass with proxy based observation as well
- Also need to add some extra tests for observation of things that weren't possible before,such as array length and index setter
- A little benchmark would be in place as well

## ⏭ Next Steps

Continue improving this and proceed with lifecycle improvements.

### CODE REVIEW

**Comment by fkleuver** [comment]
@EisenbergEffect @bigopon I would like to get some early feedback on this since this is going to have potentially huge impact on the course of the framework, and we should think carefully about how to control this via the API.
It's not a pretty nor very flexible implementation and needs many more tests. It's most likely not particularly efficient, doesn't catch everywhere yet and probably breaks in some edge cases. 
Still, at the basic level it seems to work.

**Comment by EisenbergEffect** [comment]
@fkleuver Is the idea that someone can choose which version they want on a system or per component basis depending on their startup vs. repaint speeds? Also, would be always use proxy for array length and index change, regardless of the setting? (Haven't read the code yet but wanted to just get this first to mind question out.)

File: packages/runtime-html/test/behavior-assistance.ts [hunk 1]
```diff
@@ -48,7 +48,7 @@ export function hydrateCustomElement<T>(Type: Constructable<T>, ctx: HTMLTestCon
 
   const renderingEngine = container.get(IRenderingEngine);
   const projectorLocator = container.get(IProjectorLocator);
-  element.$hydrate(dom, projectorLocator, renderingEngine, host);
+  element.$hydrate(0, dom, projectorLocator, renderingEngine, host);
```
Comment (BBosman) [inline packages/runtime-html/test/behavior-assistance.ts]: `LifecycleFlags.none` instead of `0`?

Comment (BBosman) [reply packages/runtime-html/test/behavior-assistance.ts]: I see there's more of these `0` calls in the PR. I'll leave it to just this comment instead of tagging them all individually. ;)

Comment (fkleuver) [reply packages/runtime-html/test/behavior-assistance.ts]: Yeah, there were about 800 of them to add (not all of which could be easily done with find/replace due to small differences) and I first wanted to see if it would work in the first place :) will fix this up

**Comment by fkleuver** [comment]
@EisenbergEffect The idea is that you can choose between proxy vs. descriptor observation even on a binding level. That would be a bit verbose to configure, but possible. Most folks would probably want to set it on their root-level `app` or pass it via `aurelia.app({ ..., useProxies: true })` and forget about it. But since it is controlled by lifecycle flags, it can be changed at any level.
For length and index observation the idea is that you can use either proxies, the regular length observation (so listening for mutation calls) or dirty checking. I'm still meaning to add a binding behavior to say "dirty check this". But nothing "automatically" uses proxies. It *could* be that way, but I figured it would be safer to not do that just yet.

File: packages/runtime/test/templating/view.spec.ts [hunk 2]
```diff
@@ -423,7 +423,7 @@ describe('View', () => {
       const rootTemplate: ITemplate<AuNode> = {
         renderContext: null as any,
         dom: null as any,
-        render(renderable) {
+        render( renderable) {
```
Comment (BBosman) [inline packages/runtime/test/templating/view.spec.ts]: Bonus space?

Comment (fkleuver) [reply packages/runtime/test/templating/view.spec.ts]: There was about 3h of mind-numbing test fixing, had to change a few things back and forth, did not pay attention to linting just yet :P

Comment (BBosman) [reply packages/runtime/test/templating/view.spec.ts]: As this was the only one I spotted during my visual check I'd consider that a good score for a 3 hour coding marathon ;)

Comment (fkleuver) [reply packages/runtime/test/templating/view.spec.ts]: Phew, haha

**Comment by fkleuver** [comment]
I think this needs to be done in multiple steps. I can't go much further without properly fixing up and testing the lifecycles. @bigopon can you do a sanity check on this? @BBosman do you have any findings regarding typings/linting?

**Comment by BBosman** [comment]
@fkleuver No new findings here. 👍

**Comment by fkleuver** [comment]
Thanks @BBosman 
@EisenbergEffect I'm merging this with the intent of cleaning up and improving a couple things revolving around this as soon as I'm done with some lifecycle improvements. For now it is to avoid blocking others and avoiding a merge conflict hot spot.
I asked @bigopon for a sanity check and he pointed out (which I agree with) that proxies shouldn't be flowing through the AST. There needs to be a smarter mechanism for this, potentially something with getter traps, but will have to experiment with that.

The PR in its current form does not in any way affect the existing runtime as long as `useProxies` is not specified, so even though it is in a rough-ish state, it can not really do any harm either.

**Comment by EisenbergEffect** [comment]
Sounds good @fkleuver I'm in full agreement with this approach.
